File: cli/run-ledger.clj
(ns north.run-ledger
  (:require [cheshire.core :as json]
            [clojure.java.io :as io]
            [clojure.string :as str]))

(def repo-root (some-> (or *file* (System/getProperty "babashka.file")) io/file .getCanonicalFile .getParentFile .getParentFile str))

(def contract (json/parse-string (slurp (str repo-root "/contracts/agent-run-ledger-v2.json"))))

(def version (get contract "version"))

(def wire-version (get contract "wireVersion"))

(def max-events (get-in contract ["bounds" "maxEventsPerRun"]))

(def max-event-bytes (get-in contract ["bounds" "maxCanonicalEventBytes"]))

(def max-batch-events (get-in contract ["bounds" "maxBatchEvents"]))

(def max-projection-batch-bytes (get-in contract ["bounds" "maxProjectionBatchBytes"]))

(def max-telemetry-projection-bytes (get-in contract ["bounds" "maxTelemetryProjectionBytes"]))

(def event-predicates (set (get contract "predicates")))

(def required-event-predicates #{"kind" "wire_ledger_version" "wire_version" "wire_run_id" "thread" "agent" "wire_event_id" "wire_event_sequence" "wire_event_at" "wire_event_kind" "wire_event_essential" "wire_event_json" "wire_event_sha256"})

(def identifier-pattern #"^[A-Za-z0-9][A-Za-z0-9_.:/-]{0,255}$")

(def wire-id-pattern #"^[A-Za-z0-9@][A-Za-z0-9@_.:/-]{0,255}$")

(def entity-pattern #"^@?[A-Za-z0-9][A-Za-z0-9_.:-]{0,255}$")

(def digest-pattern #"^[a-f0-9]{64}$")

(defn fail! [message data]
  (throw (ex-info message data)))

(defn- canonical-value [value]
  (cond
  (map? value) (into (sorted-map) (map (fn [[k v]] [k (canonical-value v)])) value)
  (sequential? value) (mapv canonical-value value)
  :else value))

(defn canonical-json [value]
  (json/generate-string (canonical-value value)))

(defn sha256 [value]
  (let [digest (.digest (java.security.MessageDigest/getInstance "SHA-256") (.getBytes (str value) java.nio.charset.StandardCharsets/UTF_8))]
  (apply str (map (fn [__north_anon_1] (format "%02x" (bit-and 255 __north_anon_1))) digest))))

(defn canonical-entity! [value label]
  (if (not (and (string? value) (re-matches entity-pattern value))) (do
  (fail! (str "invalid wire ledger " label) {:value value})))
  (if (str/starts-with? value "@") value (str "@" value)))

(defn event-subject [run sequence]
  (str "@run:wire-event-" (sha256 (str "north-wire-event-subject:v2\u0000" run "\u0000" sequence))))

(defn run-summary-subject! [run]
  (if (not (and (string? run) (re-matches wire-id-pattern run))) (do
  (fail! "invalid wire run id" {:value run})))
  (if (re-matches #"^run:[A-Za-z0-9][A-Za-z0-9._:-]*$" run) (str "@" run) (str "@run:wire-summary-" (sha256 (str "north-wire-run-summary-subject:v2\u0000" run)))))

(defn- singleton-map! [facts]
  (let [grouped (group-by first facts)]
  (doseq [[predicate entries] grouped]
  (if (> (count entries) 1) (do
  (fail! "wire event predicates must be singleton" {:predicate predicate :values (mapv second entries)}))))
  (into {} (map (fn [[predicate entries]] [predicate (second (first entries))])) grouped)))

(defn- parse-sequence! [value]
  (let [sequence (parse-long (or value ""))]
  (if (not (and sequence (<= 0 sequence) (< sequence max-events))) (do
  (fail! "invalid wire event sequence" {:value value})))
  sequence))

(defn- parse-instant! [value]
  (try
  (java.time.Instant/parse value)
  (catch Exception _
    (fail! "invalid wire event timestamp" {:value value}))))

(defn- parse-event-json! [raw]
  (if (not (and (string? raw) (<= (alength (.getBytes raw java.nio.charset.StandardCharsets/UTF_8)) max-event-bytes) (not (str/includes? raw "\n")) (not (str/includes? raw "\r")))) (do
  (fail! "wire event JSON is missing, multiline, or oversized" {})))
  (let [event (try
  (json/parse-string raw)
  (catch Exception error
    (fail! "invalid wire event JSON" {:cause (.getMessage error)})))]
  (if (not (map? event)) (do
  (fail! "wire event JSON must encode an object" {})))
  event))

(defn validate-event-facts! [subject facts]
  (let [unknown (seq (remove event-predicates (map first facts)))
   scalar (singleton-map! facts)
   missing (seq (remove (fn [__north_anon_1] (contains? scalar __north_anon_1)) required-event-predicates))
   raw (get scalar "wire_event_json")
   event (parse-event-json! raw)
   sequence (parse-sequence! (get scalar "wire_event_sequence"))
   run (get scalar "wire_run_id")
   thread (if (= "(ad-hoc)" (get scalar "thread")) "(ad-hoc)" (canonical-entity! (get scalar "thread") "thread"))
   agent (get scalar "agent")
   event-id (get scalar "wire_event_id")
   event-kind (get scalar "wire_event_kind")
   event-at (get scalar "wire_event_at")
   essential (get scalar "wire_event_essential")
   digest (sha256 raw)
   expected-subject (event-subject run sequence)]
  (if unknown (do
  (fail! "wire event contains unknown predicates" {:predicates unknown})))
  (if missing (do
  (fail! "wire event is missing required predicates" {:predicates missing})))
  (if (not (= "wire_event" (get scalar "kind"))) (do
  (fail! "wire event requires kind=wire_event" {})))
  (if (not (= version (get scalar "wire_ledger_version"))) (do
  (fail! "unsupported wire ledger version" {:version (get scalar "wire_ledger_version")})))
  (if (not (and (= wire-version (get scalar "wire_version")) (= wire-version (get event "version")))) (do
  (fail! "wire event version mismatch" {})))
  (if (not (and (string? agent) (re-matches identifier-pattern agent))) (do
  (fail! "invalid wire event agent" {})))
  (if (not (and (string? event-id) (re-matches wire-id-pattern event-id))) (do
  (fail! "invalid wire event id" {})))
  (if (not (and (string? run) (re-matches wire-id-pattern run))) (do
  (fail! "invalid wire run id" {})))
  (if (not (and (string? event-kind) (not (str/blank? event-kind)) (<= (count event-kind) 128))) (do
  (fail! "invalid wire event kind" {})))
  (if (not (#{"true" "false"} essential)) (do
  (fail! "invalid wire event essential flag" {})))
  (parse-instant! event-at)
  (let [bind__5 (get scalar "parent_thread")]
  (if bind__5 (let [parent-thread bind__5]
  (do
  (canonical-entity! parent-thread "parent_thread")))))
  (let [bind__6 (get scalar "run_coordinator")]
  (if bind__6 (let [coordinator bind__6]
  (do
  (if (not (some? (re-matches identifier-pattern coordinator))) (do
  (fail! "invalid wire event coordinator" {})))))))
  (if (not (and (= run (get event "runId")) (= sequence (get event "sequence")) (= event-id (get event "id")) (= event-kind (get event "kind")) (= event-at (get event "at")) (= (= essential "true") (get event "essential")))) (do
  (fail! "wire event envelope differs from its indexed facts" {})))
  (if (not (and (vector? (get event "requiredSemantics")) (every? (fn [__north_anon_1] (and (string? __north_anon_1) (not (str/blank? __north_anon_1)))) (get event "requiredSemantics")))) (do
  (fail! "wire event requiredSemantics is malformed" {})))
  (if (not (and (re-matches digest-pattern (or (get scalar "wire_event_sha256") "")) (= digest (get scalar "wire_event_sha256")))) (do
  (fail! "wire event digest mismatch" {:expected digest})))
  (if (not (= expected-subject (canonical-entity! subject "event subject"))) (do
  (fail! "wire event subject does not match run and sequence" {:expected expected-subject :actual subject})))
  {"subject" expected-subject "run" run "thread" thread "agent" agent "parentThread" (get scalar "parent_thread") "coordinator" (get scalar "run_coordinator") "sequence" sequence "id" event-id "at" event-at "kind" event-kind "essential" (= essential "true") "json" raw "digest" digest "event" event}))

(defn ledger-digest [events]
  (sha256 (canonical-json (mapv (fn [__north_anon_1] (get __north_anon_1 "digest")) events))))

(defn timeline [run-id events]
  (let [run run-id
   ordered (vec (sort-by (fn [__north_anon_1] (get __north_anon_1 "sequence")) events))
   sequences (mapv (fn [__north_anon_1] (get __north_anon_1 "sequence")) ordered)
   expected (vec (range (count ordered)))
   terminal (last ordered)]
  {:run run :thread (get (first ordered) "thread") :agent (get (first ordered) "agent") :parent-thread (get (first ordered) "parentThread") :coordinator (get (first ordered) "coordinator") :events ordered :valid-order? (= expected sequences) :finalized? (and (= "run.terminated" (get terminal "kind")) (= (dec (count ordered)) (get terminal "sequence"))) :digest (if (seq ordered) (do
  (ledger-digest ordered)))}))

File: sdk/src/run-ledger.js
import { keyword as $$bc$keyword, str as $$bc$str } from './bridge/generated/beagle/core.js';
import { admit_host_object as $$bh$admit_host_object, aset as $$bh$aset, host_object as $$bh$host_object, js_obj as $$bh$js_obj } from './bridge/generated/beagle/host.js';
import { catch_dispatch as $$bd$catch_dispatch } from './bridge/generated/beagle/exception-dispatch.js';

const node_module = process.getBuiltinModule("node:module");

const create_require = node_module.createRequire;

const require_module = create_require(import.meta.url);

const path_module = process.getBuiltinModule("node:path");

const fs_module = process.getBuiltinModule("node:fs");

const crypto_module = process.getBuiltinModule("node:crypto");

const resolve = path_module.resolve;

const read_file_sync = fs_module.readFileSync;

const create_hash = crypto_module.createHash;

const bun_spawn = Bun.spawn;

const beagle_store_module = require_module("./beagle-store");

const beagle_store_babashka_arguments = beagle_store_module.beagleStoreBabashkaArguments;

const beagle_store_environment = beagle_store_module.beagleStoreEnvironment;

const settle_beagle_store_coordinator_child = beagle_store_module.settleBeagleStoreCoordinatorChild;

const wire_events_module = require_module("./wire/events");

const WIRE_MAX_EVENTS_PER_RUN = wire_events_module.WIRE_MAX_EVENTS_PER_RUN;

const WIRE_VERSION = wire_events_module.WIRE_VERSION;

const wire_jsonl_module = require_module("./wire/jsonl");

const encode_wire_jsonl_line = wire_jsonl_module.encodeWireJsonlLine;

const wire_reducer_module = require_module("./wire/reducer");

const reduce_wire_events = wire_reducer_module.reduceWireEvents;

const RECORDED = "recorded";

const UNAVAILABLE = "unavailable";

const REPO = resolve(import.meta.dir, "../..");

const CONTRACT_PATH = resolve(REPO, "contracts/agent-run-ledger-v2.json");

const INTERNAL_WRITER = resolve(REPO, "cli/run-event-internal.clj");

const IDENTIFIER = new RegExp("^[A-Za-z0-9][A-Za-z0-9_.:/-]{0,255}$");

const WIRE_ID = new RegExp("^[A-Za-z0-9@][A-Za-z0-9@_.:/-]{0,255}$");

const ENTITY = new RegExp("^@?[A-Za-z0-9][A-Za-z0-9_.:-]{0,255}$");

const SHA256 = new RegExp("^[a-f0-9]{64}$");

const AGENT__RUN__LEDGER__CONTRACT = Object.freeze(JSON.parse(read_file_sync(CONTRACT_PATH, "utf8")));

const AGENT__RUN__LEDGER__VERSION = AGENT__RUN__LEDGER__CONTRACT.version;

function wire_ledger_error_new_bang(...$beagle$args) {
  if (arguments.length === 2) {
    const code = $beagle$args[0];
    const message = $beagle$args[1];
    return wire_ledger_error_new_bang(code, message, null);
  }
  if (arguments.length === 3) {
    const code = $beagle$args[0];
    const message = $beagle$args[1];
    const cause = $beagle$args[2];
    const error = new Error(message);
    Object.setPrototypeOf(error, wire_ledger_error_new_bang.prototype);
    (error.name = "WireLedgerError");
    (error.code = code);
    if (((_truthy) => _truthy !== false && _truthy != null)(cause)) {
      (error.cause = cause);
    }
    return error;
  }
  throw new Error('No matching arity: ' + $beagle$args.length);
}

function wire_ledger_error_constructor_bang(code, message, options) {
  return wire_ledger_error_new_bang(code, message, (((_truthy) => _truthy !== false && _truthy != null)(options) ? options.cause : null));
}

Object.setPrototypeOf(wire_ledger_error_new_bang.prototype, wire_ledger_error_constructor_bang.prototype);

Object.setPrototypeOf(wire_ledger_error_constructor_bang.prototype, Error.prototype);

const WireLedgerError = wire_ledger_error_constructor_bang;

function ledger_error_bang(...$beagle$args) {
  if (arguments.length === 2) {
    const code = $beagle$args[0];
    const message = $beagle$args[1];
    return (() => { throw wire_ledger_error_new_bang(code, message); })();
  }
  if (arguments.length === 3) {
    const code = $beagle$args[0];
    const message = $beagle$args[1];
    const cause = $beagle$args[2];
    return (() => { throw wire_ledger_error_new_bang(code, message, cause); })();
  }
  throw new Error('No matching arity: ' + $beagle$args.length);
}

function canonical_entity_bang(value, label) {
  if ((!((_truthy) => _truthy !== false && _truthy != null)(ENTITY.test(value)))) {
    ledger_error_bang("invalid_identity", $$bc$str("invalid wire ledger ", label));
  }
  return (((_truthy) => _truthy !== false && _truthy != null)(value.startsWith("@")) ? value : $$bc$str("@", value));
}

function wire_run_ledger_identity_bang(identity) {
  if ((!((_truthy) => _truthy !== false && _truthy != null)(IDENTIFIER.test(identity.agent)))) {
    ledger_error_bang("invalid_identity", "invalid wire ledger agent");
  }
  const thread = ((identity.thread === "(ad-hoc)") ? identity.thread : canonical_entity_bang(identity.thread, "thread"));
  const parent_thread = ((identity.parentThread === undefined) ? null : canonical_entity_bang(identity.parentThread, "parentThread"));
  const coordinator = ((identity.coordinator === undefined) ? null : identity.coordinator.replace(new RegExp("^@agent:"), ""));
  const result = $$bh$js_obj("thread", thread, "agent", identity.agent);
  if (((_truthy) => _truthy !== false && _truthy != null)(((_logical) => (_logical !== false && _logical != null ? (!((_truthy) => _truthy !== false && _truthy != null)(IDENTIFIER.test(coordinator))) : _logical))(coordinator))) {
    ledger_error_bang("invalid_identity", "invalid wire ledger coordinator");
  }
  if (((_truthy) => _truthy !== false && _truthy != null)(parent_thread)) {
    (($beagle$host$arg$0, $beagle$host$arg$1, $beagle$host$arg$2) => $$bh$aset($$bh$admit_host_object($beagle$host$arg$0), $beagle$host$arg$1, $beagle$host$arg$2))(result, "parentThread", parent_thread);
  }
  if (((_truthy) => _truthy !== false && _truthy != null)(coordinator)) {
    (($beagle$host$arg$0, $beagle$host$arg$1, $beagle$host$arg$2) => $$bh$aset($$bh$admit_host_object($beagle$host$arg$0), $beagle$host$arg$1, $beagle$host$arg$2))(result, "coordinator", coordinator);
  }
  return Object.freeze(result);
}

const wireRunLedgerIdentity = wire_run_ledger_identity_bang;

function sha256(value) {
  const hash = create_hash("sha256");
  hash.update(value);
  return hash.digest("hex");
}

function canonical_string_array(values) {
  return $$bc$str("[", values.map((value) => JSON.stringify(value)).join(","), "]");
}

function canonical_event_bang(event) {
  return (() => { try {
    const line = encode_wire_jsonl_line(event, $$bh$js_obj("maxLineBytes", AGENT__RUN__LEDGER__CONTRACT.bounds.maxCanonicalEventBytes));
  const canonical = $$bh$js_obj("event", event, "json", line.slice(0, -1));
  return canonical;
  } catch (_catch_0) {
    switch ($$bd$catch_dispatch(_catch_0, [Error])) {
      case 0: {
        const error = _catch_0;
        return (() => { throw wire_ledger_error_new_bang("invalid_event", "wire ledger event is invalid or oversized", error); })();
        break;
      }
    }
  } })();
}

function event_subject(run_id, sequence) {
  const digest = sha256($$bc$str("north-wire-event-subject:v2\x00", run_id, "\x00", sequence));
  return $$bc$str("@run:wire-event-", digest);
}

function fact(predicate, value) {
  const result = [predicate, value];
  return result;
}

function wire_event_facts_bang(identity, event) {
  const context = wire_run_ledger_identity_bang(identity);
  const canonical = canonical_event_bang(event);
  const digest = sha256(canonical.json);
  const facts = [fact("kind", "wire_event"), fact("wire_ledger_version", AGENT__RUN__LEDGER__VERSION), fact("wire_version", canonical.event.version), fact("wire_run_id", canonical.event.runId), fact("thread", context.thread), fact("agent", context.agent), fact("wire_event_id", canonical.event.id), fact("wire_event_sequence", String(canonical.event.sequence)), fact("wire_event_at", canonical.event.at), fact("wire_event_kind", canonical.event.kind), fact("wire_event_essential", String(canonical.event.essential)), fact("wire_event_json", canonical.json), fact("wire_event_sha256", digest)];
  if ((!(context.parentThread === undefined))) {
    facts.push(fact("parent_thread", context.parentThread));
  }
  if ((!(context.coordinator === undefined))) {
    facts.push(fact("run_coordinator", context.coordinator));
  }
  return Object.freeze($$bh$host_object($$bc$keyword("subject"), event_subject(canonical.event.runId, canonical.event.sequence), $$bc$keyword("facts"), Object.freeze(facts)));
}

const wireEventFacts = wire_event_facts_bang;

function validate_event_slice_bang(events) {
  if ((events.length === 0)) {
    ledger_error_bang("invalid_batch", "wire ledger batch must not be empty");
  }
  if ((events.length > AGENT__RUN__LEDGER__CONTRACT.bounds.maxBatchEvents)) {
    ledger_error_bang("invalid_batch", $$bc$str("wire ledger batch exceeds ", AGENT__RUN__LEDGER__CONTRACT.bounds.maxBatchEvents, " events"));
  }
  const first = events[0];
  (() => { let index = 0; while (true) {
    if ((index < events.length)) { const event = events[index]; canonical_event_bang(event); (((!(event.runId === first.runId)) || (!(event.sequence === (first.sequence + index)))) ? (() => { return ledger_error_bang("invalid_batch", "wire ledger batch must be one contiguous run slice"); })() : null); (((index < (events.length - 1)) && (event.kind === "run.terminated")) ? (() => { return ledger_error_bang("invalid_batch", "wire ledger batch cannot continue after run.terminated"); })() : null); const _recur_0 = (index + 1); index = _recur_0; continue; } else { return null; }
  } })();
  return null;
}

function terminal_event(events) {
  return events[(events.length - 1)];
}

const TERMINAL_LIFECYCLES = ["completed", "failed", "cancelled", "blocked"];

function wire_ledger_summary_bang(events) {
  if (((events.length === 0) || (events.length > WIRE_MAX_EVENTS_PER_RUN))) {
    ledger_error_bang("invalid_summary", "wire ledger summary requires a bounded event sequence");
  }
  const snapshot = (() => { try {
    return reduce_wire_events(events);
  } catch (_catch_1) {
    switch ($$bd$catch_dispatch(_catch_1, [Error])) {
      case 0: {
        const error = _catch_1;
        return (() => { throw wire_ledger_error_new_bang("invalid_summary", "wire ledger summary requires a valid event sequence", error); })();
        break;
      }
    }
  } })();
  const terminal = terminal_event(events);
  if (((!(terminal.kind === "run.terminated")) || (!((_truthy) => _truthy !== false && _truthy != null)(TERMINAL_LIFECYCLES.includes(snapshot.lifecycle))))) {
    ledger_error_bang("invalid_summary", "wire ledger summary requires run.terminated last");
  }
  const digests = events.map((event) => sha256(canonical_event_bang(event).json));
  return Object.freeze($$bh$host_object($$bc$keyword("version"), AGENT__RUN__LEDGER__VERSION, $$bc$keyword("wireVersion"), WIRE_VERSION, $$bc$keyword("runId"), snapshot.runId, $$bc$keyword("eventCount"), events.length, $$bc$keyword("firstSequence"), 0, $$bc$keyword("lastSequence"), terminal.sequence, $$bc$keyword("terminalEventId"), terminal.id, $$bc$keyword("digest"), sha256(canonical_string_array(digests))));
}

const wireLedgerSummary = wire_ledger_summary_bang;

function validate_batch_bang(events) {
  validate_event_slice_bang(events);
  const first = events[0];
  const terminal = terminal_event(events);
  if (((!(first.sequence === 0)) || ((!(first.kind === "run.started")) || (!(terminal.kind === "run.terminated"))))) {
    ledger_error_bang("invalid_batch", "wire ledger publication requires one complete terminal run");
  }
  (() => { try {
    return wire_ledger_summary_bang(events);
  } catch (_catch_2) {
    switch ($$bd$catch_dispatch(_catch_2, [Error])) {
      case 0: {
        const error = _catch_2;
        return ledger_error_bang("invalid_batch", "wire ledger publication requires a reducible terminal run", error);
        break;
      }
    }
  } })();
  return null;
}

function projection_payload_bang(projections) {
  const payload = JSON.stringify(projections);
  if ((new TextEncoder().encode(payload).byteLength > AGENT__RUN__LEDGER__CONTRACT.bounds.maxProjectionBatchBytes)) {
    ledger_error_bang("invalid_batch", "wire ledger projection batch exceeds its byte bound");
  }
  return payload;
}

async function run_writer_bang(projections, timeout_ms, env) {
  const payload = projection_payload_bang(projections);
  const command = ["bb"].concat(beagle_store_babashka_arguments([INTERNAL_WRITER, ((_logical) => (_logical !== false && _logical != null ? _logical : "7977"))(env.NORTH_PORT)], env));
  const spawn_options = $$bh$js_obj("env", beagle_store_environment(env), "stdin", "pipe", "stdout", "ignore", "stderr", "ignore");
  const child = bun_spawn(command, spawn_options);
  child.stdin.write(payload);
  child.stdin.end();
  const outcome = await settle_beagle_store_coordinator_child(child, timeout_ms);
  return Promise.resolve((((!((_truthy) => _truthy !== false && _truthy != null)(outcome.timedOut)) && (outcome.exitCode === 0)) ? RECORDED : UNAVAILABLE));
}

async function record_wire_event_projections_bang(...$beagle$args) {
  if (arguments.length === 1) {
    const projections = $beagle$args[0];
    return record_wire_event_projections_bang(projections, 10000, process.env);
  }
  if (arguments.length === 2) {
    const projections = $beagle$args[0];
    const timeout_ms = $beagle$args[1];
    return record_wire_event_projections_bang(projections, timeout_ms, process.env);
  }
  if (arguments.length === 3) {
    const projections = $beagle$args[0];
    const timeout_ms = $beagle$args[1];
    const env = $beagle$args[2];
    if ((projections.length === 0)) {
      ledger_error_bang("invalid_batch", "wire ledger projection batch must not be empty");
    }
    return (() => { try {
    return run_writer_bang(projections, timeout_ms, env);
  } catch (_catch_3) {
    switch ($$bd$catch_dispatch(_catch_3, [Error])) {
      case 0: {
        const __error = _catch_3;
        return Promise.resolve(UNAVAILABLE);
        break;
      }
    }
  } })();
  }
  throw new Error('No matching arity: ' + $beagle$args.length);
}

const recordWireEventProjections = record_wire_event_projections_bang;

function default_wire_ledger_batch_writer_bang(projections, timeout_ms) {
  return record_wire_event_projections_bang(projections, timeout_ms);
}

const DEFAULT_WIRE_LEDGER_BATCH_WRITER = default_wire_ledger_batch_writer_bang;

function event_projections_bang(identity, events) {
  return events.map((event) => wire_event_facts_bang(identity, event));
}

function create_wire_event_store_publisher_bang(...$beagle$args) {
  if (arguments.length === 1) {
    const identity = $beagle$args[0];
    return create_wire_event_store_publisher_bang(identity, $$bh$js_obj());
  }
  if (arguments.length === 2) {
    const identity = $beagle$args[0];
    const options = $beagle$args[1];
    const context = wire_run_ledger_identity_bang(identity);
    const timeout_ms = ((_logical) => (_logical !== false && _logical != null ? _logical : 10000))(options.timeoutMs);
    const writer = ((_logical) => (_logical !== false && _logical != null ? _logical : DEFAULT_WIRE_LEDGER_BATCH_WRITER))(options.writer);
    const next_sequence = ({value: 0, watches: {}});
    const poisoned = ({value: null, watches: {}});
    return $$bh$js_obj("publish", (events) => { const failure = poisoned.value;
return (((_truthy) => _truthy !== false && _truthy != null)(failure) ? Promise.reject(failure) : (() => { try {
    validate_event_slice_bang(events);
  if ((!(events[0].sequence === next_sequence.value))) {
    ledger_error_bang("invalid_batch", "wire event Store suffix does not begin at the next sequence");
  }
  const projections = event_projections_bang(context, events);
  const published = writer(projections, timeout_ms);
  const settled = published.then((status) => { if ((!(status === "recorded"))) {
  (() => { throw wire_ledger_error_new_bang("invalid_batch", "wire event Store publication is unavailable"); })();
}
(() => { const _a = next_sequence, _v = (next_sequence.value + events.length); const _old = _a.value; _a.value = _v; for (const _k in _a.watches) _a.watches[_k](_k, _a, _old, _v); return _v; })();
return null; });
  projection_payload_bang(projections);
  return settled.catch((error) => { (() => { const _a = poisoned, _v = error; const _old = _a.value; _a.value = _v; for (const _k in _a.watches) _a.watches[_k](_k, _a, _old, _v); return _v; })();
return (() => { throw error; })(); });
  } catch (_catch_4) {
    switch ($$bd$catch_dispatch(_catch_4, [Error])) {
      case 0: {
        const error = _catch_4;
        (() => { const _a = poisoned, _v = error; const _old = _a.value; _a.value = _v; for (const _k in _a.watches) _a.watches[_k](_k, _a, _old, _v); return _v; })();
        return Promise.reject(error);
        break;
      }
    }
  } })()); });
  }
  throw new Error('No matching arity: ' + $beagle$args.length);
}

const createWireEventStorePublisher = create_wire_event_store_publisher_bang;

async function publish_wire_events_bang(...$beagle$args) {
  if (arguments.length === 2) {
    const identity = $beagle$args[0];
    const events = $beagle$args[1];
    return publish_wire_events_bang(identity, events, 10000, DEFAULT_WIRE_LEDGER_BATCH_WRITER);
  }
  if (arguments.length === 3) {
    const identity = $beagle$args[0];
    const events = $beagle$args[1];
    const timeout_ms = $beagle$args[2];
    return publish_wire_events_bang(identity, events, timeout_ms, DEFAULT_WIRE_LEDGER_BATCH_WRITER);
  }
  if (arguments.length === 4) {
    const identity = $beagle$args[0];
    const events = $beagle$args[1];
    const timeout_ms = $beagle$args[2];
    const writer = $beagle$args[3];
    validate_batch_bang(events);
    const projections = event_projections_bang(identity, events);
    projection_payload_bang(projections);
    return writer(projections, timeout_ms);
  }
  throw new Error('No matching arity: ' + $beagle$args.length);
}

const publishWireEvents = publish_wire_events_bang;

function is_wire_run_ledger_summary_p(value) {
  return ((!(value == null)) && ((typeof value === "object") && ((!Array.isArray(value)) && ((value.version === AGENT__RUN__LEDGER__VERSION) && ((value.wireVersion === WIRE_VERSION) && ((typeof value.runId === "string") && ((_logical) => (_logical !== false && _logical != null ? (Number.isSafeInteger(value.eventCount) && ((value.eventCount > 0) && ((value.firstSequence === 0) && (Number.isSafeInteger(value.lastSequence) && ((value.lastSequence === (value.eventCount - 1)) && ((typeof value.terminalEventId === "string") && ((_logical) => (_logical !== false && _logical != null ? ((typeof value.digest === "string") && SHA256.test(value.digest)) : _logical))(WIRE_ID.test(value.terminalEventId)))))))) : _logical))(WIRE_ID.test(value.runId))))))));
}

const isWireRunLedgerSummary = is_wire_run_ledger_summary_p;

export { AGENT__RUN__LEDGER__CONTRACT as "AGENT_RUN_LEDGER_CONTRACT" };
export { AGENT__RUN__LEDGER__VERSION as "AGENT_RUN_LEDGER_VERSION" };
export { WireLedgerError as "WireLedgerError" };
export { createWireEventStorePublisher as "createWireEventStorePublisher" };
export { isWireRunLedgerSummary as "isWireRunLedgerSummary" };
export { publishWireEvents as "publishWireEvents" };
export { recordWireEventProjections as "recordWireEventProjections" };
export { wireEventFacts as "wireEventFacts" };
export { wireLedgerSummary as "wireLedgerSummary" };
export { wireRunLedgerIdentity as "wireRunLedgerIdentity" };
